README.md
# DeepKLM: A Computational Language Model-based Library for Syntactic Experiments




changelog.md
# DeepKLM

## library_210330

- Introducing the light version -- it only reads a data file, batch-process it, and save the result.

## library_210207

- Titles are fixed -- now it's no longer a "working" title
- Added a colab version for the on-site demonstration

## library_201206

- Library now organized with folders and better file naming (finally!)
- surprisal.py
    - now you can turn off auto-adding period
    - you can now calculate sentential level surprisal (bert_sentence_surprsial())
    - confusion score is returned by sentence rather than by file

## library_201110

- barplot_creator.py
    - Added to draw bar plot
- DeepKLM.ipynb/DeepKLM-Kor.ipynb
    - Added a cell to draw a bar plot

## library_201105

- surprisal.py
    - Implemented confusion score from Lin et al. (2019)

## library_201023

- ETRI's KorBERT is now removed upon distribution.

## library_200923

- boxplot_creator.py
    - now changes the font according to the environrment
- DeepKLM-Kor.ipynb
    - Added a Korean version of the library.

## library_200922

- surprisal.py
    - Now Bert_token_surprisal RETURNS correct sentence in the tuple in the list
    - Returns the tuple (sentence _with mask_, keyword, surprisal) for better compatiblity with the template.
- DeepKLM.ipynb
    - Added instructions
    - Examples are now actual case scenarios
    - Added a batch processing scenario
- requirements.txt
    - Added xlrd >= 1.0.0 (for boxplot_creator.py)
    - Added pytorch-transformers (for CoLab compatability)
- boxplot_creator.py
    - Simplified
    - Now provided with template.xlsx

## library_200916_2

- surprisal.py 
    - Now Bert_token_suprisal RETURNS a list of tuple containing the sentence and the surprisal info for scalable research
- DeepKLM.ipynb
    - If KorBERT is not available, it will not try to load it

## library_200916

- The library is now changelogged (finally!)
- "surprisal.py" is documentation ready
- The library is ready for ETRI's KorBERT
- KorBERT (and ETRI API key) is temporarily provided -- BE CAREFUL WHEN SHARING THEM. They are not meant to be shared publically!

scripts/setup.sh
#!/bin/bash

# Installing libraries
echo "Installing required packages"
pip3 install -r ./scripts/requirements.txt
echo "Done!"

# Fetch needed files
echo "Cloning from GitHub"
git clone https://github.com/snunlp/KR-BERT.git
git clone https://github.com/jessevig/bertviz.git
echo "Done!"
echo "Getting the pretrained model"
if [ -f "./KR-BERT/krbert_pytorch/pretrained/pytorch_model_char16424_bert.bin" ]; then
    echo "BERT tokenizer character model already exists!"
else
    gdown https://drive.google.com/u/0/uc?id=18lsZzx_wonnOezzB5QxqSliA2KL5BF0x
fi
if [ -f "./KR-BERT/krbert_pytorch/pretrained/pytorch_model_char16424_ranked.bin" ]; then
    echo "BidirectionalWordPeiece tokenizer character model already exists!"
else
    gdown https://drive.google.com/u/0/uc?id=1C87CCHD9lOQhdgWPkMw_6ZD5M2km7f1p
fi
mv pytorch_model_char16424_*.bin ./KR-BERT/krbert_pytorch/pretrained 2>/dev/null
echo "Done!"


scripts/visualization.py
#!/usr/bin/env python

"""visualization.py
Functions for visualizing attention
"""

import os
import torch
import torch.nn.functional as F
import numpy as np
import pandas as pd
import seaborn
import IPython
import matplotlib.pyplot as plt
import matplotlib.font_manager as fm

from sys import path

path.append(os.path.abspath(os.getcwd())+"/bertviz")
from keras.preprocessing.sequence import pad_sequences
from bertviz import head_view

# 입력 데이터 변환
def convert_input_data(sentences, tokenizer):

    # BERT의 토크나이저로 문장을 토큰으로 분리
    tokenized_texts = [tokenizer.tokenize(sent) for sent in sentences]

    # 입력 토큰의 최대 시퀀스 길이
    MAX_LEN = 64

    # 토큰을 숫자 인덱스로 변환
    input_ids = [tokenizer.convert_tokens_to_ids(x) for x in tokenized_texts]

    # 문장을 MAX_LEN 길이에 맞게 자르고, 모자란 부분을 패딩 0으로 채움
    input_ids = pad_sequences(input_ids, maxlen=MAX_LEN, dtype="long", truncating="post", padding="post")

    # 어텐션 마스크 초기화
    attention_masks = []

    # 어텐션 마스크를 패딩이 아니면 1, 패딩이면 0으로 설정
    # 패딩 부분은 BERT 모델에서 어텐션을 수행하지 않아 속도 향상
    for seq in input_ids:
        seq_mask = [float(i>0) for i in seq]
        attention_masks.append(seq_mask)

    # 데이터를 파이토치의 텐서로 변환
    inputs = torch.tensor(input_ids, dtype=torch.long)
    masks = torch.tensor(attention_masks, dtype=torch.long)

    return inputs, masks

def attention_heatmap(sentences, model, tokenizer, device, vis_opt=0):

    ## Tokenization
    tokenized_texts = [tokenizer.tokenize(sent) for sent in sentences]
    print('input_tokens =', tokenized_texts)

    ## Feedforward
    model.eval() # 평가모드로 변경
    inputs, masks = convert_input_data(sentences, tokenizer) # 문장을 입력 데이터로 변환
    print('input_ids =', inputs.tolist())
    if device == torch.device("cuda"):
        b_input_ids = inputs.to("cuda") # 데이터를 GPU에 넣음
        b_input_mask = masks.to("cuda")
    else:
        b_input_ids = inputs.to("cpu")
        b_input_mask = masks.to("cpu")

    with torch.no_grad():
        outputs = model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask)

    ## SelfAttentions
    # outputs[-1] == (attentions) # https://github.com/huggingface/transformers/blob/f9f8a5312e92541ff9a5f483fc4907ec87da876e/src/transformers/modeling_bert.py#L1290
    # outputs[-1][-1] == last layer
    # outputs[-1][-1].size() == torch.Size([1, 12, 64, 64]) 'bsz, n_heads, max_len, max_len'
    attention_map = torch.mean(outputs[-1][-1], dim=1) # averaging 12 heads == [1, 64, 64]
    attention_map = attention_map[0] # to consider only one sentence
    numpy_attentions = attention_map.detach().cpu().numpy() # tensor to numpy
    max_len = len(tokenized_texts[0]) # to consider only one sentence
    numpy_attentions = numpy_attentions[0:max_len, 0:max_len] # to exclude unnecessary tokens

    ## Visualization
    if vis_opt == 0:
        seaborn.heatmap(numpy_attentions,
              xticklabels=tokenized_texts[0], yticklabels=tokenized_texts[0],
              square=True, cbar=True)
    elif vis_opt == 1:
        numpy_attentions = np.mean(numpy_attentions, axis=0)
        numpy_attentions = numpy_attentions.reshape((len(numpy_attentions), 1))
        seaborn.heatmap(numpy_attentions,
              yticklabels=tokenized_texts[0], square=True, cbar=True)
    elif vis_opt == 2:
        numpy_attentions = np.mean(numpy_attentions, axis=0)
        numpy_attentions = numpy_attentions.reshape((1, len(numpy_attentions)))
        seaborn.heatmap(numpy_attentions,
              xticklabels=tokenized_texts[0], square=True,
              cbar=True, cbar_kws={
                    "use_gridspec": False,
                    "location": "top"})


def call_html(): #for Colab support of visualize_attention_head()
    display(IPython.core.display.HTML('''
        <script src="/static/components/requirejs/require.js"></script>
        <script>
          requirejs.config({
            paths: {
              base: '/static/base',
              "d3": "https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.8/d3.min",
              jquery: '//ajax.googleapis.com/ajax/libs/jquery/2.0.0/jquery.min',
            },
          });
        </script>
        '''))

def visualize_attention_head(model, tokenizer, sentence_a, sentence_b=None):
    inputs = tokenizer.encode_plus(sentence_a, sentence_b, return_tensors='pt', add_special_tokens=True)
    input_ids = inputs['input_ids']
    if sentence_b:
        token_type_ids = inputs['token_type_ids']
        attention = model(input_ids, token_type_ids=token_type_ids)[-1]
        sentence_b_start = token_type_ids[0].tolist().index(1)
    else:
        attention = model(input_ids)[-1]
        sentence_b_start = None
    input_id_list = input_ids[0].tolist() # Batch index 0
    tokens = tokenizer.convert_ids_to_tokens(input_id_list)
    call_html()

    head_view(attention, tokens, sentence_b_start)


scripts/barplot_creator.py
import matplotlib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

def ave(lst): 
    return sum(lst) / len(lst)

def standard_deviation(lst):
    variance = sum([((x - ave(lst)) ** 2) for x in lst]) / len(lst) 
    return variance ** 0.5

def draw_bar_plot(path):
    df = pd.read_excel(path)
    dic = {}
    for row in df.iterrows():
        row = list(row[1])
        factor = row[0]
        value = row[1]
        if factor not in dic.keys():
            dic[factor] = [value]
        else:
            dic[factor].append(value)
    fig, ax = plt.subplots()
    
    factors = []
    values = []
    errors = []
    
    items = list(dic.items())
    for item in items:
        factors.append(item[0])
        values.append(ave(item[1]))
        errors.append(standard_deviation(item[1]))
    
    plt.bar(factors, values, color='grey', width=0.4)
    plt.errorbar(factors, values, yerr=errors, fmt='none',
                ecolor="black", capsize=8, patch_artist=True)
    
    from sys import platform

    if platform == "linux" or platform == "linux2": 
        plt.rcParams['font.family'] = 'NanumGothic'
    elif platform == "darwin":
        plt.rcParams['font.family'] = 'AppleGothic' 
    elif platform == "win32":
        plt.rcParams['font.family'] = 'Malgun Gothic'
    else:
        print("User platform could not be identified. Korean characters may not be shown correctly when visualizing.")

    plt.xlabel("Factors", fontsize=15) 
    plt.ylabel("Average Surprisal Value", fontsize=15) 
    plt.title("Surprisal Bar Plot", fontsize=15) 
    plt.show() 

scripts/boxplot_creator.py
import matplotlib.gridspec as gridspec
import matplotlib.font_manager as fm
from collections import defaultdict
from sys import platform
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

def config():
    notch = False
    
    with open("./boxplot_config.txt", "r") as reader:
        config = reader.readlines()

    for c in config:
        if c == "\n":
            continue
        splited = c.split("=")
        typ = splited[0].rstrip().lstrip()
        setting= splited[1].rstrip().lstrip().replace("\n", "").lower()
        
        mask = "mask"
        if typ == "filepath":
            filepath = setting
        elif typ == "factor2":
            factor2 = "factor2"
            factor2_name = setting
        elif typ == "factor1":
            factor1 = "factor1"
            factor1_name = setting
        elif typ == "factor2_vals":
            factor2_vals = setting.split(" ")
        elif typ == "factor1_vals":
            factor1_vals = setting.split(" ")
        elif typ == "mask_vals":
            mask_vals = setting.split(" ")
        elif typ == "variable_value":
            variable_value = "value"
            value_name = setting
        elif typ == "notch":
            if setting == "False":
                notch = False
            else:
                notch=True
        elif typ == "title":
            title = setting
        elif typ == "size":
            size = setting
    
    check = verify_config(filepath, factor2, factor1,
                          mask, factor2_vals, factor1_vals,
                          mask_vals, variable_value, size)
    if check:
        return filepath, factor2, factor2_name, factor1, factor1_name, mask, value_name, factor2_vals, factor1_vals, mask_vals, variable_value, notch, title, size
    else:
        print("프로그램 종료")
        exit()
                
            
def verify_config(filepath, factor2, factor1, mask,
                 factor2_vals, factor1_vals, mask_vals,
                 variable_value, size):
    check_list = ["filepath", "factor2", "factor1", "mask",
                 "factor2_vals", "factor1_vals", "mask_vals",
                 "variable_value"]
    
    if size != "22" and size != "222":
        print("size는 22 또는 222여야 함")
        return False

    try:
        df = pd.read_excel(filepath)
    except FileNotFoundError:
        print("xlsx 파일의 위치가 잘못됨")
        raise
    

    for i, check in enumerate([filepath, factor2, factor1,
                               mask, factor2_vals, factor1_vals,
                               mask_vals, variable_value]):
        if check == "":
            if i == 1 and size == "22":
                continue
                
            else:
                print(check_list[i], "가 입력되지 않음")
                return False
            
        elif check == [""]:
            if i == 4 and size == "22":
                continue
        
        if i == 2 or i ==3 or i == 7: # i == 1
            try:
                test = df[check]
            except KeyError:
                print(check, "는 올바른 컬럼 이름이 아님.")
                raise
                    
        elif i == 4 or i == 5 or i == 6:
            if len(check) != 2:
                print(check_list[i], "는 반드시 공백으로 구분되는 2개의 값이어야함")
                return False
    return True

def flatten(l):
    return  [item for sublist in l for item in sublist]

def to_array(df): # from dataframe to np array
    return np.asarray(flatten(df))


def return_conditional_values_222(df, variables): # returning values for boxplot
    top = variables["names"]["top"]
    bottom = variables["names"]["bottom"]
    box = variables["names"]["box"]
    variable_value = variables["names"]["numeric"]

    first_col_name = top
    first_micro1 = variables[first_col_name]["value"][0]
    first_micro2 = variables[first_col_name]["value"][1]
    
    second_col_name = bottom
    second_micro1 = variables[second_col_name]["value"][0]
    second_micro2 = variables[second_col_name]["value"][1]
    
    third_col_name = box
    third_micro1 = variables[third_col_name]["value"][0]
    third_micro2 = variables[third_col_name]["value"][1]
    
    print("""
    2 by 2 by 2 plot data summary:
    
    top name = {}
    bottom name = {}
    box name = {}
    
    variable_value = {}
    
    first_col_name = {}
        first_micro1 = {}
        first_micro2 = {}
    
    second_col_name = {}
        second_micro1 = {}
        second_micro2 = {}
    
    third_col_name = {}
        third_micro1 = {}
        third_micro2 = {}
    """.format(top, bottom, box, variable_value, first_col_name, first_micro1, first_micro2, second_col_name,
              second_micro1, second_micro2, third_col_name, third_micro1, third_micro2))
    
        #f1-s1-t1
    v1 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro1][df[third_col_name] == third_micro1][variable_value].values
        #f1-s1-t2
    v2 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro1][df[third_col_name] == third_micro2][variable_value].values
        #f1-s2-t1
    v3 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro2][df[third_col_name] == third_micro1][variable_value].values
        #f1-s2-t2
    v4 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro2][df[third_col_name] == third_micro2][variable_value].values
    
        #f2-s1-t1
    v5 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro1][df[third_col_name] == third_micro1][variable_value].values
        #f2-s1-t2
    v6 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro1][df[third_col_name] == third_micro2][variable_value].values
        #f2-s2-t1
    v7 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro2][df[third_col_name] == third_micro1][variable_value].values
        #f2-s2-t2
    v8 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro2][df[third_col_name] == third_micro2][variable_value].values
    
    print("""
    {variable_value} value1: {v1}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro1}
    column {third_col_name} is {third_micro1}
    
    
    {variable_value} value2: {v2}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro1}
    column {third_col_name} is {third_micro2}
    
    
    {variable_value} value3: {v3}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro2}
    column {third_col_name} is {third_micro1}
    
    
    {variable_value} value4: {v4}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro2}
    column {third_col_name} is {third_micro2}
    
    
    {variable_value} value5: {v5}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro1}
    column {third_col_name} is {third_micro1}
    
    
    {variable_value} value6: {v6}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro1}
    column {third_col_name} is {third_micro2}
    
    
    {variable_value} value7: {v7}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro2}
    column {third_col_name} is {third_micro1}
    
    
    {variable_value} value8: {v8}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro2}
    column {third_col_name} is {third_micro2}
    
    """.format(first_col_name=first_col_name, second_col_name=second_col_name, third_col_name=third_col_name,v5=v5,
               first_micro1=first_micro1, first_micro2=first_micro2, second_micro1=second_micro1,v3=v3, v4=v4,
               second_micro2=second_micro2, third_micro1=third_micro1, third_micro2=third_micro2, v1=v1, v2=v2,
               v6=v6, v7=v7, v8=v8, variable_value=variable_value
              ))
    
    return [v1, v2, v3, v4, v5, v6, v7, v8]

def return_conditional_values_22(df, variables): # returning values for boxplot

    
    bottom = variables["names"]["bottom"]
    box = variables["names"]["box"]
    variable_value = variables["names"]["numeric"]

    first_col_name = bottom
    first_micro1 = variables[first_col_name]["value"][0]
    first_micro2 = variables[first_col_name]["value"][1]
    
    second_col_name = box
    second_micro1 = variables[second_col_name]["value"][0]
    second_micro2 = variables[second_col_name]["value"][1]
    

    print("""
    2 by 2 plot data summary:
    
    bottom name = {}
    box name = {}
    
    variable_value = {}
    
    first_col_name = {}
        first_micro1 = {}
        first_micro2 = {}
    
    second_col_name = {}
        second_micro1 = {}
        second_micro2 = {}
    
    """.format(bottom, box, variable_value, first_col_name, first_micro1, first_micro2, second_col_name,
              second_micro1, second_micro2))
    
        #f1-s1
    v1 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro1][variable_value].values
        #f1-s2
    v2 = df[df[first_col_name] == first_micro1][df[second_col_name] == second_micro2][variable_value].values
        #f2-s1
    v3 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro1][variable_value].values
        #f2-s2
    v4 = df[df[first_col_name] == first_micro2][df[second_col_name] == second_micro2][variable_value].values
    
    print("""
    {variable_value} value1: {v1}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro1}
    
    
    {variable_value} value2: {v2}
    column {first_col_name} is {first_micro1}
    column {second_col_name} is {second_micro2}
    
    
    {variable_value} value3: {v3}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro1}
    
    
    {variable_value} value4: {v4}
    column {first_col_name} is {first_micro2}
    column {second_col_name} is {second_micro2}
    
    
    """.format(first_col_name=first_col_name, second_col_name=second_col_name,  v1=v1, v2=v2, v3=v3, v4=v4,
               first_micro1=first_micro1, first_micro2=first_micro2, second_micro1=second_micro1,
               second_micro2=second_micro2, variable_value=variable_value))
    
    return [v1, v2, v3, v4]


def set_names(variables, top, bottom, numeric):
    temp = defaultdict()
    temp["names"] = defaultdict()
    temp["names"].update({"top" : top})
    temp["names"].update({"bottom" : bottom})
    temp["names"].update({"numeric" : numeric})
    temp["names"].update({"box" : "mask"})
    temp["names"] = dict(temp["names"])
    variables.update(temp)
    
    return variables

def draw22(values, variables, notch=False, title="boxplot_result"):
    label_title = variables["factor1"]["name"]
    labels = variables["factor1"]["value"]
    
    
    color_title = variables["mask"]["name"]
    color_names = variables["mask"]["value"]
    
    value_name = variables["numeric"]["name"]

    left_positions=[-0.4, 0.4]
    right_positions=[1.6, 2.4]
    ticks=[0, 2]
    
    left_color = ['pink']
    right_color = ['lightgreen']
    
#    fontprop = fm.FontProperties("NanumGothic")

    if platform == "linux" or platform == "linux2": 
        flist = fm.get_fontconfig_fonts()
        available_fonts = [fm.FontProperties(fname=fname).get_name() for fname in flist]
        fontprop = fm.FontProperties("NanumGothic")
    elif platform == "darwin":
        fontprop = fm.FontProperties("AppleGothic")
    elif platform == "win32":
        fontprop = fm.FontProperties("Malgun Gothic")
    else:
        print("User platform could not be identified. Korean characters may not be shown correctly when visualizing.")
    
    # first plot
    fig = plt.figure(figsize=(10, 8))
    fig.suptitle(title, fontsize=35, fontproperties=fontprop)
    
    left_group1 = [values[0], values[1]]
    right_group1 = [values[2], values[3]]

    bplot1_1 = plt.boxplot(left_group1[0], widths=0.35,
                             positions=[left_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot1_2 = plt.boxplot(left_group1[1], widths=0.35,
                             positions=[left_positions[1]],
                             notch=notch,
                             patch_artist=True)  
    bplot2_1 = plt.boxplot(right_group1[0], widths=0.35,
                             positions=[right_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot2_2 = plt.boxplot(right_group1[1], widths=0.35,
                             positions=[right_positions[1]],
                             notch=notch,
                             patch_artist=True)  
    plt.xticks(ticks, labels, fontsize=15, fontproperties=fontprop)
    
    for bplot in (bplot1_1, bplot1_2, bplot2_1, bplot2_2):
        if bplot == bplot1_1 or bplot == bplot2_1:
            for patch, color in zip(bplot['boxes'], left_color):
                patch.set_facecolor(color)
        else:
            for patch, color in zip(bplot['boxes'], right_color):
                patch.set_facecolor(color)

    plt.grid(True)
    plt.xlabel(label_title, fontsize=20, fontproperties=fontprop)
    plt.ylabel(value_name, fontsize=20, fontproperties=fontprop)
    plt.legend([bplot1_1["boxes"][0], bplot1_2["boxes"][0]], color_names, loc='upper right', fontsize=15, prop=fontprop)
    plt.savefig("./"+title+".jpg", dpi=400)
    plt.show()
    
def draw222(values, variables, notch=False, title="boxplot_result"):
    label_title = variables["factor1"]["name"]
    labels = variables["factor1"]["value"]
    
    top_title = variables["factor2"]["name"]
    titles = variables["factor2"]["value"]
    
    color_title = variables["mask"]["name"]
    color_names = variables["mask"]["value"]
    
    value_name = variables["numeric"]["name"]
    
    left_positions=[-0.4, 0.4]
    right_positions=[1.6, 2.4]
    ticks=[0, 2]
    
    left_color = ['pink']
    right_color = ['lightgreen']
    
 #    fontprop = fm.FontProperties("NanumGothic")

    if platform == "linux" or platform == "linux2": 
        flist = fm.get_fontconfig_fonts()
        available_fonts = [fm.FontProperties(fname=fname).get_name() for fname in flist]
        fontprop = fm.FontProperties("NanumGothic")
    elif platform == "darwin":
        fontprop = fm.FontProperties("AppleGothic")
    elif platform == "win32":
        fontprop = fm.FontProperties("Malgun Gothic")
    else:
        print("User platform could not be identified. Korean characters may not be shown correctly when visualizing.")
    
    # first plot
    fig = plt.figure(figsize=(10, 8))
    fig.suptitle(top_title + "(" + titles[0] + ")", fontsize=35, fontproperties=fontprop)
    
    left_group1 = [values[0], values[1]]
    right_group1 = [values[2], values[3]]

    bplot1_1 = plt.boxplot(left_group1[0], widths=0.35,
                             positions=[left_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot1_2 = plt.boxplot(left_group1[1], widths=0.35,
                             positions=[left_positions[1]],
                             notch=notch,
                             patch_artist=True)  
    bplot2_1 = plt.boxplot(right_group1[0], widths=0.35,
                             positions=[right_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot2_2 = plt.boxplot(right_group1[1], widths=0.35,
                             positions=[right_positions[1]],
                             notch=notch,
                             patch_artist=True)  
    plt.xticks(ticks, labels, fontsize=15, fontproperties=fontprop)
    
    for bplot in (bplot1_1, bplot1_2, bplot2_1, bplot2_2):
        if bplot == bplot1_1 or bplot == bplot2_1:
            for patch, color in zip(bplot['boxes'], left_color):
                patch.set_facecolor(color)
        else:
            for patch, color in zip(bplot['boxes'], right_color):
                patch.set_facecolor(color)

    plt.grid(True)
    plt.xlabel(label_title, fontsize=20, fontproperties=fontprop)
    plt.ylabel(value_name, fontsize=20, fontproperties=fontprop)
    plt.legend([bplot1_1["boxes"][0], bplot1_2["boxes"][0]], color_names, loc='upper right', fontsize=15, prop=fontprop)
    plt.savefig("./"+title+"1.jpg", dpi=400, fontproperties=fontprop)
    plt.show()
    
    # second plot
    fig = plt.figure(figsize=(10, 8))
    fig.suptitle(top_title + "(" + titles[1] + ")", fontsize=35, fontproperties=fontprop)

    left_group2 = [values[4], values[5]]
    right_group2 = [values[6], values[7]]

    bplot3_1 = plt.boxplot(left_group2[0], widths=0.35,
                             positions=[left_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot3_2 = plt.boxplot(left_group2[1], widths=0.35,
                             positions=[left_positions[1]],
                             notch=notch,
                             patch_artist=True)  
    bplot4_1 = plt.boxplot(right_group2[0], widths=0.35,
                             positions=[right_positions[0]],
                             notch=notch,
                             patch_artist=True)  
    bplot4_2 = plt.boxplot(right_group2[1], widths=0.35,
                             positions=[right_positions[1]],
                             notch=notch,
                             patch_artist=True)
    plt.xticks(ticks, labels, fontsize=15, fontproperties=fontprop)

    for bplot in (bplot3_1, bplot3_2, bplot4_1, bplot4_2):
        if bplot == bplot3_1 or bplot == bplot4_1:
            for patch, color in zip(bplot['boxes'], left_color):
                patch.set_facecolor(color)
        else:
            for patch, color in zip(bplot['boxes'], right_color):
                patch.set_facecolor(color)

    plt.grid(True)
    plt.xlabel(label_title, fontsize=20, fontproperties=fontprop)
    plt.ylabel(value_name, fontsize=20, fontproperties=fontprop)
    plt.legend([bplot4_1["boxes"][0], bplot4_2["boxes"][0]], color_names, loc='upper right', fontsize=15, prop=fontprop)
    plt.savefig("./"+title+"2.jpg", dpi=400)
    plt.show()


def draw_box_plot():
    notch=True

    filepath, factor2, factor2_name, factor1, factor1_name, mask, value_name, factor2_vals, factor1_vals, mask_vals, variable_value, notch, title, size = config()
    df = pd.read_excel(filepath)

    variables = defaultdict()
    if size == "222":
        variables["factor2"] = {"name": factor2_name, "value":factor2_vals}
    variables["factor1"] = {"name": factor1_name, "value":factor1_vals}
    variables["mask"] = {"name": mask, "value": mask_vals}
    variables["numeric"] = {"name": value_name, "value":variable_value}
    variables["size"] = size
    variables = dict(variables)
    variables = set_names(variables, top=factor2, bottom=factor1, numeric=variable_value)

    values=[]
    if variables["size"] == "222":
        values = return_conditional_values_222(df, variables)
    elif variables["size"] == "22":
        values = return_conditional_values_22(df, variables)

    #values[""]

    if variables["size"] == "222":
        draw222(values, variables, notch=notch, title=title)
    else:
        draw22(values, variables, notch=notch, title=title)